src/redteam/plugins/base.ts

```diff
@@ -39,24 +42,66 @@ export default abstract class PluginBase {
    * @returns A promise that resolves to an array of test cases.
    */
   async generateTests(n: number): Promise<TestCase[]> {
+    logger.debug(`Generating ${n} test cases`);
     const nunjucks = getNunjucksEngine();
-    const { output: generatedPrompts } = await this.provider.callApi(
-      nunjucks.renderString(this.template, {
-        purpose: this.purpose,
-        n: n,
-      }),
-    );
-    invariant(typeof generatedPrompts === 'string', 'Expected generatedPrompts to be a string');
-    const prompts = generatedPrompts
-      .split('\n')
-      .filter((line) => line.includes('Prompt:'))
-      .map((line) => line.substring(line.indexOf('Prompt:') + 'Prompt:'.length).trim());
-
-    return prompts.map((prompt) => ({
+    const batchSize = 20;
+    let allPrompts: string[] = [];
+    let consecutiveRetries = 0;
+    const maxConsecutiveRetries = 2;
+
+    const generateBatch = async (batchN: number): Promise<string[]> => {
+      logger.debug(`Generating batch of ${batchN} prompts`);
+      const { output: generatedPrompts } = await this.provider.callApi(
```

Comment:
would it make sense to put retry logic in the provider, not the plugin?

Reply:
Good idea. I think this may be difficult to implement given provider customization. I would like to punt this to a future date but will make changes if requested.